conans/client/tools/settings.py

```diff
@@ -1,77 +1,382 @@
+                                        "recipe is missing")
+    try:
+        if compiler not in required:
+            raise ConanInvalidConfiguration("Compiler support information is missing")
+    except ConanInvalidConfiguration as e:
+        if strict:
+            raise
+        conanfile.output.warn(e)
+        return
+
+    compiler_version = conanfile.settings.get_safe("compiler.version")
+    if not compiler_version:
+        raise ConanInvalidConfiguration("Could not obtain cppstd because compiler.version "
+                                        "is not specified in the profile")
+    version = Version(compiler_version)
+    if version < required[compiler]:
+        raise ConanInvalidConfiguration(
+            "At least {} {} is required".format(compiler, required[compiler]))
+
+
+def compatible_cppstd(conanfile, current_cppstd=None, min=None, max=None,
+                      backward=False, forward=True, gnu_extensions_compatible=True):
+    """ Fill the conanfile compatible_packages field with compatible cppstd
+
+    :param conanfile: ConanFile instance
+    :param current_cppstd: Current cppstd to override deduction
+    :param min: Inclusive cppstd lower bound
+    :param max: Inclusive cppstd upper bound
+    :param backward: Current cppstd is compatible with lower ones
+    :param forward: Current cppstd is compatible with higher ones
+    :param gnu_extensions_compatible: GNU extensions are not compatible
+    """
+    if current_cppstd is not None:
+        if not isinstance(current_cppstd, str) and not str(current_cppstd).isdigit():
+            raise ConanException("current_cppstd parameter must either be a string or a digit")
+    if min is not None and not str(min).isdigit():
+        raise ConanException("min parameter must be a number")
+    if max is not None and not str(max).isdigit():
+        raise ConanException("max parameter must be a number")
+    if not isinstance(backward, bool):
+        raise ConanException("backward parameter must be a bool")
+    if not isinstance(forward, bool):
+        raise ConanException("forward parameter must be a bool")
+    if not isinstance(gnu_extensions_compatible, bool):
+        raise ConanException("gnu_extensions_compatible parameter must be a bool")
+
+    # We allow overriding cppstd deduction to leave a place for
+    # custom user settings
+    if not current_cppstd:
+        current_cppstd = deduced_cppstd(conanfile)
+        if not current_cppstd:
+            raise ConanInvalidConfiguration(
+                "Could not detect default cppstd for the current compiler.")
+    else:
+        current_cppstd = str(current_cppstd)
+
+    # We have a list of original cppstd setting values because we'll
+    # need it to pass to compatible packages
+    # FIXME: Conanfile should carry range information of a setting with it
+    base_cppstd_range = ["98", "11", "14", "17", "20"]
```

Comment:
can we share it from settings? For now we have init file, which contains the yaml content, but we could move the cppstd list and re-use it for setting.yml and here.

Reply:
That's exactly what my comment in the `FIXME` suggests, but it sounds like it'd be better to do it in a separate PR.

Reply:
To elaborate, one of the numerous structures that can carry settings actually contains the range field which contains all the values we'd need for a given setting. The problem is that it's used inconsistently across the codebase, so it requires a cleanup.